From c323d6f25ecf6fa33522ac4603b9d899ac36d316 Mon Sep 17 00:00:00 2001
From: John Andersen <john.s.andersen@intel.com>
Date: Tue, 30 Jun 2020 19:21:19 -0700
Subject: [PATCH 4/6] x86/setup: Add kernel parameter to disable kexec

Allow for disabling kexec via a dedicated command line parameter which
happens before setting of sysctls. This is needed for control register
pinning due to it requiring an understanding of if kexec should be
disabled before kexec_load_disabled would be set via a sysctl.

Signed-off-by: John Andersen <john.s.andersen@intel.com>
---
 Documentation/admin-guide/kernel-parameters.txt | 5 +++++
 arch/x86/kernel/setup.c                         | 6 ++++++
 2 files changed, 11 insertions(+)

diff --git a/Documentation/admin-guide/kernel-parameters.txt b/Documentation/admin-guide/kernel-parameters.txt
index 89386f6f3ab6..45b231d3f4e8 100644
--- a/Documentation/admin-guide/kernel-parameters.txt
+++ b/Documentation/admin-guide/kernel-parameters.txt
@@ -917,6 +917,11 @@
 	disable_ipv6=	[IPV6]
 			See Documentation/networking/ipv6.rst.
 
+	disable_kexec	[KNL,X86] Disable kexec. Due to it's reliance on kexec,
+			crashkernel will be ignored if this option is given.
+			Specifying this option sets the kexec_load_disabled
+			sysctl to 1.
+
 	disable_mtrr_cleanup [X86]
 			The kernel tries to adjust MTRR layout from continuous
 			to discrete, to make X server driver able to add WB
diff --git a/arch/x86/kernel/setup.c b/arch/x86/kernel/setup.c
index d9c678b37a9b..99e44fbc7f33 100644
--- a/arch/x86/kernel/setup.c
+++ b/arch/x86/kernel/setup.c
@@ -484,6 +484,12 @@ static void __init reserve_crashkernel(void)
 	bool high = false;
 	int ret;
 
+	if (cmdline_find_option_bool(boot_command_line, "disable_kexec")) {
+		kexec_load_disabled = 1;
+		pr_info("Ignoring crashkernel due to disable_kexec\n");
+		return;
+	}
+
 	total_mem = memblock_phys_mem_size();
 
 	/* crashkernel=XM */
-- 
2.21.0

